device/cuda/src/seeding2/seed_finding.cu

```diff
@@ -0,0 +1,108 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include <cooperative_groups.h>
+
+#include <limits>
+#include <memory>
+#include <traccc/cuda/seeding2/kernels/kd_tree_kernel.hpp>
+#include <traccc/cuda/seeding2/kernels/seed_finding_kernel.hpp>
+#include <traccc/cuda/seeding2/kernels/write_output_kernel.hpp>
+#include <traccc/cuda/seeding2/seed_finding.hpp>
+#include <traccc/cuda/seeding2/types/internal_seed.hpp>
+#include <traccc/cuda/seeding2/types/kd_tree.hpp>
+#include <traccc/cuda/utils/definitions.hpp>
+#include <traccc/edm/seed.hpp>
+#include <traccc/edm/spacepoint.hpp>
+#include <traccc/seeding/detail/lin_circle.hpp>
+#include <traccc/seeding/detail/seeding_config.hpp>
+#include <vecmem/memory/memory_resource.hpp>
+#include <vecmem/memory/unique_ptr.hpp>
+#include <vector>
+
+namespace traccc::cuda {
+std::tuple<vecmem::unique_alloc_ptr<internal_spacepoint<spacepoint>[]>,
+           uint32_t>
+copy_sps_to_device(vecmem::memory_resource& mr,
+                   const spacepoint_container_types::const_view& sps) {
+    std::chrono::high_resolution_clock::time_point time_ex_start =
+        std::chrono::high_resolution_clock::now();
+
+    const traccc::spacepoint_container_types::const_device p(sps);
+
+    std::vector<internal_spacepoint<spacepoint>> spacepoints;
+
+    for (std::size_t i = 0; i < p.get_items().size(); ++i) {
+        for (std::size_t j = 0; j < p.get_items().at(i).size(); ++j) {
+            const traccc::spacepoint& sp = p.get_items().at(i).at(j);
+
+            spacepoints.push_back(
+                internal_spacepoint<spacepoint>(p, {i, j}, {0.f, 0.f}));
+        }
+    }
+
+    uint32_t n_sp = spacepoints.size();
+
+    // Move the spacepoints to device memory.
+    vecmem::unique_alloc_ptr<internal_spacepoint<spacepoint>[]>
+        spacepoints_device =
+            vecmem::make_unique_alloc<internal_spacepoint<spacepoint>[]>(
+                mr, n_sp, spacepoints.data(),
+                [](internal_spacepoint<spacepoint>* dst,
+                   const internal_spacepoint<spacepoint>* src,
+                   std::size_t bytes) {
+                    CUDA_ERROR_CHECK(
+                        cudaMemcpy(dst, src, bytes, cudaMemcpyHostToDevice));
+                });
+
+    return {std::move(spacepoints_device), n_sp};
```

Comment:
Why not a `vecmem::data::vector_buffer<traccc::internal_spacepoint<traccc::spacepoint> >`? Then you wouldn't need to invent a custom data type here.

Reply:
Not sure I understand the comment here, this is using vecmem datatypes, nothing new?

Reply:
You're passing an "owning pointer" and a size in an `std::tuple` here. Instead of just using `vecmem::data::vector_buffer`, which is exactly meant for this type of data storage...